src/client/api/v1/get_balance.c

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdio.h>
+#include <string.h>
+
+#include "client/api/json_utils.h"
+#include "client/api/v1/get_balance.h"
+#include "client/api/v1/response_error.h"
+#include "client/network/http.h"
+#include "core/utils/iota_str.h"
+
+res_balance_t *res_balance_new() {
+  res_balance_t *res = malloc(sizeof(res_balance_t));
+  res->u.output_balance = malloc(sizeof(get_balance_t));
+  res->is_error = false;
+  return res;
+}
+
+void res_balance_free(res_balance_t *res) {
+  if (res) {
+    if (res->is_error) {
+      res_err_free(res->u.error);
+    } else {
+      free(res->u.output_balance);
+    }
+    free(res);
+  }
+}
+
+int deser_balance_info(char const *const j_str, res_balance_t *res) {
+  char const *const key_addr = "address";
+  char const *const key_maxResults = "maxResults";
+  char const *const key_count = "count";
+  char const *const key_balance = "balance";
+  char const *const key_code = "code";
+  char const *const key_message = "message";
+  int ret = 0;
+
+  cJSON *json_obj = cJSON_Parse(j_str);
+  if (json_obj == NULL) {
+    return -1;
+  }
+
+  res_err_t *res_err = deser_error(json_obj);
+  if (res_err) {
+    // got an error response
+    res->is_error = true;
+    res->u.error = res_err;
```

Comment:
I partially address this issue on dda044c94471b60cf005b1f53df4895df88d182e

I was thinking about doing the following, please tell me your feedback:
```
   if (res_err) {
     // got an error response
     res->is_error = true;
-    res->u.error = res_err;
+    res->u.error = calloc(1, sizeof(res_err_t));
+    memcpy(res->u.error, res_err, sizeof(res_err_t));
+    free(res_err);
     ret = -1;
     goto end;
   }

```

Reply:
you don't need to allocate memory for `res->u.error` here.
I think the logic is simple here, `deser_error` parsing the response object to find if there is an error. if it's an error, the `deser_error` allocates memory, puts error data into the `res_err` object, and assigns it to `res->u.error`, otherwise `res_err` is NULL.

Reply:
ok I think I understand, please correct me if I'm wrong:

the problem was caused because I was doing `malloc` for `res->u.output_balance` inside `res_balance_new()`.
but then in case of error memory would leak from union.

to fix the problem, it's just a matter of allocating `res->u.output_balance` during `deser_balance_info` call, after we already know there is no error.

that should have been fixed via dda044c94471b60cf005b1f53df4895df88d182e, although line `res->is_error = false;` could probably remain inside `res_balance_new()`.

correct?